Adding support for prefixed transition properties

doctyper opened the pull request:
Addresses issue #289.

The transition mixin now adds appropriate prefixes for both transform and transform-origin properties. The mixin is also built to handle any additional prefixed properties the W3C throws at us down the line.

@chriseppstein I built a couple of helper functions for array / string manipulation. Let me know if I'm duplicating any already-existing functionality.

doctyper committed:
Adding support for prefixed transition properties, e.g. transform & transform-origin. Can easily be extended for further properties.

scottdavis commented:
/cc @chriseppstein @paulirish

paulirish commented:
Yessssss

Richard i love you

I think you should add in the official unprefixed version here. 

 also it's not place to ask but I imagine it'd be nice to have a test written for this feature.

doctyper committed:
Updated with the following:

- Added stylesheet tests for prefixed transition properties.
- transition-property mixin now passes all tests.
- single-transition mixin now passes all tests.
- Updated rewrite to respect user specified browser support.

doctyper commented:
@paulirish The unprefixed version is covered using the [@experimental mixin](https://github.com/chriseppstein/compass/pull/728/files#L0L130). I just landed test cases for the changes as well. Good thing too, since some of my previous changes were apparently broken.

Also, not sure if you guys agree but my changes are now on the beefy side. @chriseppstein is there anything I can do to slim this down?

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -78,29 +100,56 @@ $default-transition-delay: false !default;
+    @if $property == $prefixed-property {
+      $match: true;
+    }
+  }
```
Setting `$match` should be doable using functions instead of loops. But it's not -- we need better list and string functions. >_<

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -23,15 +23,37 @@ $default-transition-function: false !default;
+    @if $experimental-support-for-webkit    { -webkit-transition-property: -webkit-#{$properties}; }
+    @if $experimental-support-for-mozilla   { -moz-transition-property: -moz-#{$properties}; }
+    @if $experimental-support-for-microsoft { -ms-transition-property: -ms-#{$properties}; }
+    @if $experimental-support-for-opera     { -o-transition-property: -o-#{$properties}; }
```
pretty sure this fails if $properties is a list. We need to map the prefix onto each of the values of the list.

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -23,15 +23,37 @@ $default-transition-function: false !default;
+    @if $experimental-support-for-opera     { -o-transition-property: -o-#{$properties}; }
+    @include experimental(transition-property, $properties,
+      not -moz, not -webkit, not -o, not -ms, not -khtml, official
+    );
```
There's really no point in calling the `experimental` mixin here if you're just setting the official version.

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -23,15 +23,37 @@ $default-transition-function: false !default;
+    @if $properties == $prefixed-property {
+      $match: true;
+    }
+  }
```
Setting $match should be doable using functions instead of loops. But it's not -- we need better list and string functions. >_<

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -78,29 +100,56 @@ $default-transition-delay: false !default;
+        @if $experimental-support-for-opera     { -o-transition: -o-#{$property} $duration $function $delay; }
+        @include experimental(transition, $property $duration $function $delay,
+          not -moz, not -webkit, not -o, not -ms, not -khtml, official
+        );
```
Again, no point in calling `experimental`

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -78,29 +100,56 @@ $default-transition-delay: false !default;
+        @if $experimental-support-for-opera     { -o-transition: -o-#{$property} $duration $function; }
+        @include experimental(transition, $property $duration $function,
+          not -moz, not -webkit, not -o, not -ms, not -khtml, official
+        );
```
Again, no point in calling experimental

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -126,8 +175,84 @@ $default-transition-delay: false !default;
+  }
+
+  @return $existing;
+}
```
This actually returns a list -- not a string. Seems like we can use `compact(join($existing, $addition))` which are already built into compass and sass.

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -126,8 +175,84 @@ $default-transition-delay: false !default;
+  }
+
+  @return $array;
+}
```
Seems the sass `append` function should suffice. Am I missing something?

chriseppstein replied:
- `append($list, $value, comma)`
- `append($list, $value, space)`
- `append($list, $value, auto)`

doctyper replied:
The issue here is that I need to append values to an initially-false variable:

```
$list: false;
append($list, foo, comma);
```

This results in:

```
$list: false, foo;
```

I expect:

```
$list: foo;
```

The only workaround I've come up with is to set the initial $list value to an unquoted string. Unfortunately, this adds an empty space to the beginning of the value and fails during testing:

```
$list: unquote("");
append($list, foo, comma);
```

Result:

```
$list:  foo;
```

Expected (sans space):

```
$list: foo;
```

Any ideas?

chriseppstein replied:
The compass `compact` function will remove false values from a list.

doctyper committed:
Using transition property in lieu of @experimental mixin.

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -126,8 +175,84 @@ $default-transition-delay: false !default;
+}
+
+@mixin prefix-friendly-transition($transition) {
+  $raw-transition: -compass-space-list($transition);
```
we don't need to use `-compass-space-list` anymore. Sass provides `()` to construct an empty space-delimited list.

chriseppstein replied:
Oh. sorry. This coerces a possible string into a list. Seems like the string case shouldn't exist anymore. But I'm foggy on the deprecation status.

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -126,8 +175,84 @@ $default-transition-delay: false !default;
+  $o-transition: false;
+
+  @each $single-transition in $raw-transition {
+    $property: nth($single-transition, 1);
```
I don't understand this. are there cases where we have a list of lists?

chriseppstein replied:
And if so, why, and is it correct to discard all but the first element?

doctyper replied:
The transition property allows for multiple transition declarations:

```
transition: transform 3s ease-out, opacity 2s ease-in;
```

Here I'm iterating through each transition and manipulating the first value in the list (transform, opacity, etc).

chriseppstein replied:
gotcha.

chriseppstein commented on frameworks/compass/stylesheets/compass/css3/_transition.scss:
```diff
@@ -126,8 +175,84 @@ $default-transition-delay: false !default;
+      @if $property == $prefixed-property {
+        $match: true;
+      }
+    }
```
3 times! MUST BE REFACTORED

doctyper replied:
Agreed, and done.

doctyper committed:
transition-property mixin now includes support for multiple prefixed values.

doctyper committed:
Test for multiple transition properties.

doctyper committed:
Redundant use of a built-in helper function. Removing string-join() for compact(join()).

doctyper committed:
Per @chriseppstein: we don't need to use `-compass-space-list` anymore. Sass provides `()` to construct an empty space-delimited list.

doctyper committed:
Redundant use of a built-in helper function. Removing array-push() for append().

doctyper committed:
Adding test for multiple transitions.

doctyper committed:
Keeping things DRY, refactoring some shared functionality. All prefix fixes now route through an @build-prefix-values mixin.

doctyper committed:
Adding a single transform transition stylesheet test.

doctyper commented:
Just pushed an update to address most/all of your concerns @chriseppstein. Important bits:
- transform-property now takes multiple properties
- removed redundant helper mixins
- refactored code to DRY things up

chriseppstein commented:
Thanks will review again soon.
